python/dgl/nn/pytorch/linear.py

```diff
@@ -0,0 +1,178 @@
+    torch.Size([100, 64])
+    """
+    def __init__(self, in_size, out_size, num_types,
+                 regularizer=None, num_bases=None):
+        super().__init__()
+        self.in_size = in_size
+        self.out_size = out_size
+        self.num_types = num_types
+        if regularizer is None:
+            self.W = nn.Parameter(torch.Tensor(num_types, in_size, out_size))
+        elif regularizer == 'basis':
+            if num_bases is None:
+                raise ValueError('Missing "num_bases" for basis regularization.')
+            self.W = nn.Parameter(torch.Tensor(num_bases, in_size, out_size))
+            self.coeff = nn.Parameter(torch.Tensor(num_types, num_bases))
+            self.num_bases = num_bases
+        elif regularizer == 'bdd':
+            if in_size % num_bases != 0 or out_size % num_bases != 0:
+                raise ValueError(
+                    'Input and output sizes must be a multiplier of num_bases.'
+                )
+            self.submat_in = in_size // num_bases
+            self.submat_out = out_size // num_bases
+            self.W = nn.Parameter(torch.Tensor(
+                num_types, num_bases * self.submat_in * self.submat_out))
+            self.num_bases = num_bases
+        else:
+            raise ValueError(
+                f'Supported regularizer options: "basis", "bdd", but got {regularizer}')
+        self.regularizer = regularizer
+        self.reset_parameters()
+
+    def reset_parameters(self):
+        """Reset parameters"""
+        if self.regularizer is None:
+            nn.init.xavier_uniform_(self.W, gain=nn.init.calculate_gain('relu'))
+        elif self.regularizer == 'basis':
+            nn.init.xavier_uniform_(self.W, gain=nn.init.calculate_gain('relu'))
+            nn.init.xavier_uniform_(self.coeff, gain=nn.init.calculate_gain('relu'))
+        elif self.regularizer == 'bdd':
+            nn.init.xavier_uniform_(self.W, gain=nn.init.calculate_gain('relu'))
+        else:
+            raise ValueError(
+                f'Supported regularizer options: "basis", "bdd", but got {regularizer}')
+
+    def get_weight(self):
+        """Get type-wise weight"""
+        if self.regularizer is None:
+            return self.W
+        elif self.regularizer == 'basis':
+            W = self.W.view(self.num_bases, self.in_size * self.out_size)
+            return (self.coeff @ W).view(self.num_types, self.in_size, self.out_size)
+        elif self.regularizer == 'bdd':
+            return self.W
+        else:
+            raise ValueError(
+                f'Supported regularizer options: "basis", "bdd", but got {regularizer}')
+
+    def forward(self, x, x_type, sorted_by_type=False):
+        """Forward computation.
+
+        Parameters
+        ----------
+        x : torch.Tensor
+            A 2D input tensor. Shape: (N, D1)
+        x_type : torch.Tensor
+            A 1D integer tensor. Each element is the corresponding type of ``x``. Shape: (N,)
```

Comment:
"Each element is the corresponding type of ``x``." -> "Type of elements in ``x`` with one-to-one correspondence."